File: modules/nostr-native/ios/NostrNativeModule.swift
import ExpoModulesCore
import Foundation

// Swift twin of android/…/NostrNativeModule.kt (Stage 2 M3 of #1036) over the
// same rust-nostr 0.44 core — UniFFI-generated bindings in Generated/
// NostrSDK.swift + the vendored nostr_sdkFFI.xcframework (fetched by
// scripts/fetch-nostr-sdk-swift.mjs; see the podspec).
// Sync `Function`s on purpose: every JS call site is synchronous (see
// unwrapWrapNsec's "stays synchronous" note); the Rust lib is statically
// linked here, so unlike Android there is no JNA/dlopen cost for warmUp to
// pay — it survives as the facade's routing latch (nativeReady).
public class NostrNativeModule: Module {
  private static let hexChars = Set("0123456789abcdef".utf8)

  // Single-entry parsed-key caches, same rationale as the Kotlin module: the
  // app signs/decrypts with the SAME viewer key thousands of times per
  // session; counterparty keys are ephemeral so caching those would not pay.
  // Locked (unlike Kotlin's unsynchronized vars) because the sync crypto
  // Functions run on the JS thread while engineStart/engineStop touch the
  // caches from the module's async task context.
  private let cacheLock = NSLock()
  private var cachedSecret: (hex: String, key: SecretKey)?
  private var cachedKeys: (hex: String, keys: Keys)?

  private lazy var engine = NostrEngine { [weak self] name, body in
    self?.sendEvent(name, body)
  }

  private func clearKeyCaches() {
    cacheLock.lock()
    defer { cacheLock.unlock() }
    cachedSecret = nil
    cachedKeys = nil
  }

  private func isHex(_ value: String, length: Int) -> Bool {
    let bytes = Array(value.utf8)
    return bytes.count == length && bytes.allSatisfy { Self.hexChars.contains($0) }
  }

  private func requireHex64(_ value: String, _ what: String) throws {
    guard isHex(value, length: 64) else {
      throw NostrInputException("\(what) must be 64 lowercase hex chars")
    }
  }

  private func secretKey(_ hex: String) throws -> SecretKey {
    try requireHex64(hex, "secretKey")
    cacheLock.lock()
    defer { cacheLock.unlock() }
    if let cached = cachedSecret, cached.hex == hex { return cached.key }
    let key = try SecretKey.parse(secretKey: hex)
    cachedSecret = (hex, key)
    return key
  }

  private func keys(_ secretKeyHex: String) throws -> Keys {
    try requireHex64(secretKeyHex, "secretKey")
    cacheLock.lock()
    defer { cacheLock.unlock() }
    if let cached = cachedKeys, cached.hex == secretKeyHex { return cached.keys }
    let parsed = try Keys.parse(secretKey: secretKeyHex)
    cachedKeys = (secretKeyHex, parsed)
    return parsed
  }

  // Inputs are pre-validated lowercase hex, so the nibble math is total.
  private func hexToData(_ hex: String) -> Data {
    let bytes = Array(hex.utf8)
    var out = Data(capacity: bytes.count / 2)
    var i = 0
    while i < bytes.count {
      let hi = bytes[i] <= 0x39 ? bytes[i] - 0x30 : bytes[i] - 0x61 + 10
      let lo = bytes[i + 1] <= 0x39 ? bytes[i + 1] - 0x30 : bytes[i + 1] - 0x61 + 10
      out.append(hi << 4 | lo)
      i += 2
    }
    return out
  }

  public func definition() -> ModuleDefinition {
    Name("NostrNative")

    // Relay-engine event stream (Stage 2 M2 contract): batched plaintext
    // rumors and a debounced reconnect signal. See NostrEngine.swift.
    // Module-qualified: the generated rust-nostr bindings define an `Events`
    // TYPE (event collection) in this same pod target, which would otherwise
    // shadow Expo's Events(...) DSL function.
    ExpoModulesCore.Events("onEngineRumorBatch", "onEngineReconnect")

    OnDestroy {
      // Dev-client reloads recreate the module — never leave a pool (or key
      // material) running behind a dead JS context.
      let engine = self.engine
      Task { [weak self] in
        await engine.stop()
        self?.clearKeyCaches()
      }
    }

    // No .so/JNA load to front-run on iOS (static linkage), but the facade
    // gates routing on warmUp resolving true — keep the contract, and return
    // false (never reject) if the Rust core is somehow unusable.
    AsyncFunction("warmUp") { () -> Bool in
      _ = SecretKey.generate()
      return true
    }

    Function("nip44Encrypt") { (secretKeyHex: String, counterpartyPubkeyHex: String, plaintext: String) -> String in
      try self.requireHex64(counterpartyPubkeyHex, "counterpartyPubkey")
      return try nip44Encrypt(
        secretKey: self.secretKey(secretKeyHex),
        publicKey: PublicKey.parse(publicKey: counterpartyPubkeyHex),
        content: plaintext,
        version: Nip44Version.v2
      )
    }

    Function("nip44Decrypt") { (secretKeyHex: String, counterpartyPubkeyHex: String, payload: String) -> String in
      try self.requireHex64(counterpartyPubkeyHex, "counterpartyPubkey")
      return try nip44Decrypt(
        secretKey: self.secretKey(secretKeyHex),
        publicKey: PublicKey.parse(publicKey: counterpartyPubkeyHex),
        payload: payload
      )
    }

    Function("schnorrSign") { (messageHashHex: String, secretKeyHex: String) -> String in
      try self.requireHex64(messageHashHex, "messageHash")
      return try self.keys(secretKeyHex).signSchnorr(message: self.hexToData(messageHashHex))
    }

    // rust-nostr's FFI exports no raw BIP-340 verify — only
    // Event.verifySignature(), which schnorr-verifies `sig` over the *stored*
    // `id` field (hash recompute is the separate verifyId()). A minimal
    // synthetic event whose id is the message hash is therefore an exact raw
    // verify; inputs are strictly hex-validated so the template cannot
    // inject JSON. Same trick, byte-for-byte, as the Kotlin module.
    Function("schnorrVerify") { (signatureHex: String, messageHashHex: String, publicKeyHex: String) -> Bool in
      guard self.isHex(signatureHex, length: 128) else {
        throw NostrInputException("signature must be 128 lowercase hex chars")
      }
      try self.requireHex64(messageHashHex, "messageHash")
      try self.requireHex64(publicKeyHex, "publicKey")
      let json =
        "{\"id\":\"\(messageHashHex)\",\"pubkey\":\"\(publicKeyHex)\",\"created_at\":0,\"kind\":1,\"tags\":[],\"content\":\"\",\"sig\":\"\(signatureHex)\"}"
      return try Event.fromJson(json: json).verifySignature()
    }

    // --- Relay engine (Stage 2 M2 contract) --------------------------------
    // nsec-only: the NIP-59 unwrap needs the secret key in-process, so JS
    // only starts the engine for the local-key signer. The key routes through
    // the same single-entry cache as the crypto functions above and is
    // cleared by engineStop.

    AsyncFunction("engineStart") { (relays: [String], viewerPubkeyHex: String, privkeyHex: String) async throws -> Bool in
      do {
        try self.requireHex64(viewerPubkeyHex, "viewerPubkey")
        try await self.engine.start(
          relays: relays,
          viewerPubkeyHex: viewerPubkeyHex,
          keys: self.keys(privkeyHex)
        )
        return true
      } catch {
        throw EngineStartException(String(describing: error))
      }
    }

    AsyncFunction("engineSubscribeWraps") { (filterJson: String, knownWrapIds: [String]) async throws -> String in
      do {
        return try await self.engine.subscribeWraps(filterJson: filterJson, seedKnownWrapIds: knownWrapIds)
      } catch {
        throw EngineSubscribeException(String(describing: error))
      }
    }

    AsyncFunction("engineStop") { () async -> Void in
      // best-effort — the key-cache clear must run regardless of engine state
      await self.engine.stop()
      self.clearKeyCaches()
    }
  }
}

internal final class NostrInputException: GenericException<String>, @unchecked Sendable {
  override var code: String { "ERR_NOSTR_INPUT" }
  override var reason: String { param }
}

internal final class EngineStartException: GenericException<String>, @unchecked Sendable {
  override var code: String { "ERR_ENGINE_START" }
  override var reason: String { param }
}

internal final class EngineSubscribeException: GenericException<String>, @unchecked Sendable {
  override var code: String { "ERR_ENGINE_SUBSCRIBE" }
  override var reason: String { param }
}

File: modules/nostr-native/android/src/main/java/com/lightningpiggy/nostrnative/NostrNativeModule.kt
package com.lightningpiggy.nostrnative

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import rust.nostr.sdk.Event
import rust.nostr.sdk.Keys
import rust.nostr.sdk.Nip44Version
import rust.nostr.sdk.PublicKey
import rust.nostr.sdk.SecretKey
import rust.nostr.sdk.nip44Decrypt
import rust.nostr.sdk.nip44Encrypt

/**
 * Thin sync surface over rust-nostr's published UniFFI bindings
 * (org.rust-nostr:nostr-sdk-kmp-android — prebuilt libnostr_sdk_ffi.so,
 * no local Rust toolchain). Stage 2 M1 of the native Nostr pipeline
 * (#1036): only the four hot primitives the src/services/nostrCrypto.ts
 * facade needs.
 *
 * Sync `Function` (not AsyncFunction) on purpose: every JS call site is
 * synchronous (see unwrapWrapNsec's "stays synchronous" note), and a
 * native op is expected to run 10–50x faster than the pure-JS @noble
 * equivalent it replaces — a synchronous JSI hop keeps the semantics
 * identical while avoiding promise-scheduling overhead per op. The
 * one-time JNA + .so load (~tens of ms) is paid by `warmUp`, which runs
 * on the module's background queue.
 */
class NostrNativeModule : Module() {
  companion object {
    private val HEX_64 = Regex("^[0-9a-f]{64}$")
    private val HEX_128 = Regex("^[0-9a-f]{128}$")
  }

  // Single-entry parsed-key caches. The app decrypts/signs with the SAME
  // viewer secret key thousands of times per session; re-parsing per call
  // would pay an FFI allocation + hex parse each op. Counterparty pubkeys
  // rotate per gift wrap (ephemeral), so caching those would not pay.
  private var cachedSecret: Pair<String, SecretKey>? = null
  private var cachedKeys: Pair<String, Keys>? = null

  private fun requireHex64(value: String, what: String): String {
    if (!HEX_64.matches(value)) {
      throw CodedException("ERR_NOSTR_INPUT", "$what must be 64 lowercase hex chars", null)
    }
    return value
  }

  private fun secretKey(hex: String): SecretKey {
    requireHex64(hex, "secretKey")
    cachedSecret?.let { (cachedHex, key) -> if (cachedHex == hex) return key }
    val key = SecretKey.parse(hex)
    cachedSecret = hex to key
    return key
  }

  private fun keys(secretKeyHex: String): Keys {
    requireHex64(secretKeyHex, "secretKey")
    cachedKeys?.let { (cachedHex, k) -> if (cachedHex == secretKeyHex) return k }
    val k = Keys(SecretKey.parse(secretKeyHex))
    cachedKeys = secretKeyHex to k
    return k
  }

  private fun hexToBytes(hex: String): ByteArray =
    ByteArray(hex.length / 2) { i ->
      ((Character.digit(hex[2 * i], 16) shl 4) or Character.digit(hex[2 * i + 1], 16)).toByte()
    }

  override fun definition() = ModuleDefinition {
    Name("NostrNative")

    // Forces the JNA classloading + libnostr_sdk_ffi.so dlopen on the
    // module's background queue so the first sync crypto call on the JS
    // thread doesn't pay it. Returns false (instead of rejecting) when the
    // native lib can't load — the facade treats that as "stay on JS".
    AsyncFunction("warmUp") { ->
      try {
        SecretKey.generate()
        true
      } catch (t: Throwable) {
        false
      }
    }

    Function("nip44Encrypt") { secretKeyHex: String, counterpartyPubkeyHex: String, plaintext: String ->
      requireHex64(counterpartyPubkeyHex, "counterpartyPubkey")
      nip44Encrypt(
        secretKey(secretKeyHex),
        PublicKey.parse(counterpartyPubkeyHex),
        plaintext,
        Nip44Version.V2,
      )
    }

    Function("nip44Decrypt") { secretKeyHex: String, counterpartyPubkeyHex: String, payload: String ->
      requireHex64(counterpartyPubkeyHex, "counterpartyPubkey")
      nip44Decrypt(
        secretKey(secretKeyHex),
        PublicKey.parse(counterpartyPubkeyHex),
        payload,
      )
    }

    Function("schnorrSign") { messageHashHex: String, secretKeyHex: String ->
      requireHex64(messageHashHex, "messageHash")
      keys(secretKeyHex).signSchnorr(hexToBytes(messageHashHex))
    }

    // rust-nostr's FFI does not export a raw BIP-340 verify — only
    // Event.verifySignature(), which schnorr-verifies `sig` over the
    // *stored* `id` field (it does NOT recompute the hash; that's the
    // separate verifyId()). So a minimal synthetic event whose id is the
    // message hash is an exact raw verify. Inputs are strictly validated
    // as hex above, so the string template cannot inject JSON.
    Function("schnorrVerify") { signatureHex: String, messageHashHex: String, publicKeyHex: String ->
      if (!HEX_128.matches(signatureHex)) {
        throw CodedException("ERR_NOSTR_INPUT", "signature must be 128 lowercase hex chars", null)
      }
      requireHex64(messageHashHex, "messageHash")
      requireHex64(publicKeyHex, "publicKey")
      val json =
        """{"id":"$messageHashHex","pubkey":"$publicKeyHex","created_at":0,"kind":1,"tags":[],"content":"","sig":"$signatureHex"}"""
      Event.fromJson(json).verifySignature()
    }
  }
}
